### ISSUE TITLE
[Exception] On creating new plan / the given data was invalid

### ISSUE THREAD

**Issue by Omranic** [rinvex/subscriptions#20]
Reported by: *Almeida Silva*

```php
$plan = app('rinvex.subscriptions.plan')->create([
            'name' => 'Pro',
            'description' => 'Pro plan',
            'price' => 9.99,
            'invoice_period' => 1,
            'invoice_interval' => 'month',
            'trial_period' => 15,
            'trial_interval' => 'd',
            'sort_order' => 1,
            'currency' => 'USD',
        ]);
```

```
.Illuminate\Support\MessageBag {#732
  #messages: array:2 [
    "signup_fee" => array:1 [
      0 => "The signup fee field is required."
    ]
    "invoice_interval" => array:1 [
      0 => "The selected invoice interval is invalid."
    ]
  ]
  #format: ":message"
}
```

**Comment by ppalmeida** [comment]
Just to make sure this is some dumb mistake made by be, this is the test method:

```
    /** @test */
    public function a_user_can_subscribe_to_a_plan()
    {
        // Create a plan and a user
        $user = factory(\App\User::class)->create([
            "email" => "john@test.com",
        ]);

        $plan = app('rinvex.subscriptions.plan')->create([
            'name' => 'Pro',
            'description' => 'Pro plan',
            'price' => 9.99,
            'invoice_period' => 1,
            'invoice_interval' => 'month',
            'trial_period' => 15,
            'trial_interval' => 'd',
            'sort_order' => 1,
            'currency' => 'USD',
        ]);

        $plan->features()->saveMany([
            new PlanFeature(['name' => 'listings', 'value' => 50, 'sort_order' => 1]),
            new PlanFeature(['name' => 'pictures_per_listing', 'value' => 10, 'sort_order' => 5]),
            new PlanFeature(['name' => 'listing_duration_days', 'value' => 30, 'sort_order' => 10, 'resettable_period' => 1, 'resettable_interval' => 'month']),
            new PlanFeature(['name' => 'listing_title_bold', 'value' => 'Y', 'sort_order' => 15])
        ]);

        // User makes a subscription:
        $user->newSubscription("Pro", $plan);

        // Assertions:
        $this->assertEquals(1, $plan->subscriptions->count());
    }
```

**Comment by noxify** [comment]
Can you try:

```
[
            'name' => 'Pro',
            'description' => 'Pro plan',
            'price' => 9.99,
            'signup_fee' => 0.00,
            'invoice_period' => 1,
            'invoice_interval' => 'm',
            'trial_period' => 15,
            'trial_interval' => 'd',
            'sort_order' => 1,
            'currency' => 'USD',
]
```

instead of:

```
[
            'name' => 'Pro',
            'description' => 'Pro plan',
            'price' => 9.99,
            'invoice_period' => 1,
            'invoice_interval' => 'month',
            'trial_period' => 15,
            'trial_interval' => 'd',
            'sort_order' => 1,
            'currency' => 'USD',
]
```

invoice interval definition:
https://github.com/rinvex/subscriptions/blob/develop/src/Models/Plan.php#L188

signup fee definition:
https://github.com/rinvex/subscriptions/blob/develop/src/Models/Plan.php#L183

**Comment by ppalmeida** [comment]
Hi, @noxify.
Thank you for your help. But it didn't work (or we just change the error, wich is good :-)

Now the error bag shows this:
```php
Illuminate\Support\MessageBag {#722
  #messages: array:1 [
    "resettable_interval" => array:1 [
      0 => "The selected resettable interval is invalid."
    ]
  ]
  #format: ":message"
}
```
And that's strange because the create_plan_features_table migration is like this:
```php
$table->char('resettable_interval', 1)->default('m');
```
My method now looks like this:

```php
/** @test */
public function a_user_can_subscribe_to_plan() {
    $user = factory(User::class)->create(["email" => "john@test.com"]);
    // $user = User::where("email", "john@test.com")->first();
    $plan = $this->_createAPlan();

    $response = $this->post(route("plans.subscribe"), [
        "user_id" => $user->id,
        "plan_id" => $plan->id,
    ]);

    $response->assertStatus(201);

}

private function _createAPlan() {
    $plan = app('rinvex.subscriptions.plan')->create([
        'name' => 'Pro',
        'description' => 'Pro plan',
        'price' => 9.99,
        'signup_fee' => 0.00,
        'invoice_period' => 1,
        'invoice_interval' => 'm',
        'trial_period' => 15,
        'trial_interval' => 'd',
        'sort_order' => 1,
        'currency' => 'BRL',
    ]);

    $plan->features()->saveMany([
        new PlanFeature(['name' => 'listings', 'value' => 50, 'sort_order' => 1]),
        new PlanFeature(['name' => 'pictures_per_listing', 'value' => 10, 'sort_order' => 5]),
        new PlanFeature(['name' => 'listing_duration_days', 'value' => 30, 'sort_order' => 10, 'resettable_period' => 1, 'resettable_interval' => 'month']),
        new PlanFeature(['name' => 'listing_title_bold', 'value' => 'Y', 'sort_order' => 15])
    ]);

    return $plan;
}
```

Any help is very appreciate. This package looks really nice, I am sure it's just simple details to make it work.

Thank you all for your help and hard work :-)

**Comment by Omranic** [comment]
All `interval` fields are now string-based that accepts (`hour`, `day`, `week`, `month`) .. You should be able to use the latest release without any issues bf8b67beaa89fbc54c6ea6a7720fab0e3dc5d8cc

**Comment by ppalmeida** [comment]
Hi, @noxify 

I got the latest version of ```dev-develop``` branch and now, when I run the code, it give these errors:

```
Illuminate\Support\MessageBag {#689
  #messages: array:2 [
    "trial_interval" => array:1 [
      0 => "The selected trial interval is invalid."
    ]
    "invoice_interval" => array:1 [
      0 => "The selected invoice interval is invalid."
    ]
  ]
  #format: ":message"
}
```

And with PlanFeature:
```
Illuminate\Support\MessageBag {#710
  #messages: array:1 [
    "resettable_interval" => array:1 [
      0 => "The selected resettable interval is invalid."
    ]
  ]
  #format: ":message"
}
array:8 [
  "plan_id" => "required|integer|exists:plans,id"
  "slug" => "required|alpha_dash|max:150|unique:plan_features,slug"
  "name" => "required|string|max:150"
  "description" => "nullable|string|max:10000"
  "value" => "required|string"
  "resettable_period" => "sometimes|integer"
  "resettable_interval" => "sometimes|string|in:hour,day,week,month"
  "sort_order" => "nullable|integer|max:10000000"
]
```
So, If I use "d" or "m" as periods, the rules are against it. Could you please, recheck them?

For example: in Plan Model (https://github.com/rinvex/subscriptions/blob/77e3f025915549b6329f6f0ffbc321a208678fa2/src/Models/Plan.php#L177) the rules are validating "month", "day" etc.

Thank you.

**Comment by Omranic** [comment]
@ppalmeida you can **NOT** use short intervals any more. That said above already:

All `interval` fields are now string-based that accepts (`hour`, `day`, `week`, `month`) .. You should be able to use the latest release without any issues bf8b67beaa89fbc54c6ea6a7720fab0e3dc5d8cc

So would you please try with the long intervals 🙂

**Comment by ppalmeida** [comment]
Hi, @Omranic.

What you said above is very strange. Because I get this error from my tests:
```php
Method Illuminate\Support\Carbon::addMs does not exist
```

And in the Period.php class I can see this line:
```
$this->interval = in_array($interval, ['d', 'w', 'm', 'y']) ? $interval : 'm';
if ($count > 0) {
    $this->period = $count;
}
$method = 'add'.ucfirst($this->interval).'s';
```

And this is exactly what I think is giving the erros. Because my factory looks like this:
```

$factory->define(Plan::class, function (Faker $faker) {
    return [
        'title' => 'Pro',
        'description' => "pro plan", // $faker->sentence($nbWords = 6, $variableNbWords = true),
        'price' => 9.99, // $faker->randomFloat($nbMaxDecimals = 2, $min = 9.0, $max = 50.0),
        'signup_fee' => 0.00,
        'invoice_period' => 1,
        'invoice_interval' => 'month',
        'trial_period' => 15,
        'trial_interval' => 'day',
        'sort_order' => 1,
        'currency' => 'BRL',
    ];
});

$factory->define(PlanFeature::class, function (Faker $faker) {
    return [
        'title' => 'listing_duration_days',
        'value' => 30,
        'sort_order' => 10,
        'resettable_period' => 1,
        'resettable_interval' => 'month'
    ];
});
```

And the errors still happening, because of this IF:
https://github.com/rinvex/subscriptions/blob/f3d51ff17639c45ee0864ea1158111335bc35d87/src/Services/Period.php#L58

**Comment by Omranic** [comment]
Nice catch, thank you @ppalmeida 👍 
That was a mistakenly leftover, and now fixed on 285f08eaec97048b4fec58da51ef8c083fdf5847 😉
